src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactoryHelpers.cs

```diff
@@ -73,6 +73,45 @@ internal static class BindingFactoryHelpers
             }
         }
 
+        // Get the "core" TMEssage type that's consistent with an IAsyncCollector pattern. 
+        // This must be in sync with the rules in BindCollector<>.
+        // Or null if hte parameter type is not consistent with IASyncCollector.
+        public static Type GetAsyncCollectorCoreType(Type parameterType)
+        {
+            // IAsyncCollector<T>
+            if (parameterType.IsGenericType)
+            {
+                var genericType = parameterType.GetGenericTypeDefinition();
+                var elementType = parameterType.GetGenericArguments()[0];
+
+                if (genericType == typeof(IAsyncCollector<>))
+                {
+                    return elementType;
+                }
+                else if (genericType == typeof(ICollector<>))
+                {
+                    return elementType;
+                }
+
+                return null;
+            }
+            else
+            {
+                if (parameterType.IsByRef)
```

Comment:
Should you check IsByRef first? I'd expect something like 'out GenericType<SomeType>' to return GenericType<SomeType> as the Type.

I *think* this would actually return null for that case.